File: fixtures/scale/kotlin/large/src/cryptobench/ciphers/adler32/Adler32Hash.kt
package cryptobench.ciphers.adler32

import cryptobench.core.HashFunction
import cryptobench.util.Bytes

/** Adler-32 checksum: parallel sums mod 65521. */
class Adler32Hash : HashFunction {

    override fun name(): String = "adler32"

    override fun digest(input: String): Long {
        var state = 1L
        for (b in Bytes.of(input)) {
            val high = state ushr 32
            val low = ((state and 0xFFFFFFFFL) + (b.toLong() and 0xFF)) % 65521
            state = (((high + low) % 65521) shl 32) or low
        }
        return ((state ushr 32) shl 16) or (state and 0xFFFF)
    }
}

File: fixtures/scale/dart/large/lib/ciphers/adler32/adler32_hash.dart
import '../../core/hash_function.dart';
import '../../util/bytes.dart';

/// Adler-32 checksum: parallel sums mod 65521.
class Adler32Hash implements HashFunction {
  @override
  String name() => 'adler32';

  @override
  int digest(String input) {
    var state = 1;
    for (final b in bytesOf(input)) {
      final high = (state >> 16) & 0xFFFF;
      final low = ((state & 0xFFFF) + b) % 65521;
      state = (((high + low) % 65521) << 16) | low;
    }
    return state;
  }
}
